libqtile/bar.py

```diff
@@ -175,34 +177,50 @@ def __init__(self, widgets, size, **config):
         self.size_calculated = 0
         self._configured = False
 
+        if isinstance(self.margin, int):
+            self.margin = [self.margin] * 4
+
+        if isinstance(self.border, int):
+            self.border = [self.border] * 4
+
         self.queued_draws = 0
 
     def _configure(self, qtile, screen):
         Gap._configure(self, qtile, screen)
 
-        if self.margin:
-            if isinstance(self.margin, int):
-                self.margin = [self.margin] * 4
+        if sum(self.margin) or sum(self.border):
+
+            if isinstance(self.border_colour, str):
+                self.border_colour = [self.border_colour] * 4
+
+            # Increase the margin size for the border. The border will be drawn
+            # in this space so the empty space will just be the margin.
+            self.margin = [m + b for m, b in zip(self.margin, self.border)]
+
             if self.horizontal:
-                self.x += self.margin[3]
-                self.width -= self.margin[1] + self.margin[3]
+                self.x += self.margin[3] - self.border[3]
+                self.width -= (self.margin[1] + self.margin[3])
                 self.length = self.width
                 if self.size == self.initial_size:
-                    self.size += self.margin[0] + self.margin[2]
+                    self.size += (self.margin[0] + self.margin[2])
                 if self.screen.top is self:
-                    self.y += self.margin[0]
+                    self.y += self.margin[0] - self.border[0]
                 else:
-                    self.y -= self.margin[2]
+                    self.y -= self.margin[2] + self.border[2]
+
             else:
-                self.y += self.margin[0]
-                self.height -= self.margin[0] + self.margin[2]
+                self.y += self.margin[0] - self.border[0]
+                self.height -= (self.margin[0] + self.margin[2])
                 self.length = self.height
-                self.size += self.margin[1] + self.margin[3]
+                self.size += (self.margin[1] + self.margin[3])
                 if self.screen.left is self:
                     self.x += self.margin[3]
                 else:
                     self.x -= self.margin[1]
 
+        width = self.width + (self.border[1] + self.border[3])
+        height = self.height + (self.border[0] + self.border[2])
```

Comment:
These brackets seem out of place :stuck_out_tongue:

Reply:
Um... yeah. Not sure what I was thinking there.